packages/jazz-tools/src/schema/subscribe.ts

```diff
@@ -0,0 +1,290 @@
+import { LocalNode, RawCoMap } from "cojson";
+import { cojsonInternals } from "cojson";
+import { Account } from "../exports.js";
+import { activeAccountContext } from "../implementation/activeAccountContext.js";
+import {
+  AnonymousJazzAgent,
+  ID,
+  RefsToResolve,
+  RefsToResolveStrict,
+  Resolved,
+  SchemaInit,
+  isRefEncoded,
+} from "../internal.js";
+import { CoMapInstance, CoMapSchema, CoMapSchemaClass } from "./coMap.js";
+
+type SubscribeListener<S extends CoMapSchema, R extends RefsToResolve<S>> = (
+  value: Resolved<S, R>,
+  unsubscribe: () => void,
+) => void;
+
+function createResolvablePromise<T>() {
+  let resolve!: (value: T) => void;
+  let reject!: (reason?: any) => void;
+
+  const promise = new Promise<T>((res, rej) => {
+    resolve = res;
+    reject = rej;
+  });
+
+  return { promise, resolve, reject };
+}
+
+type ResolvablePromise<T> = ReturnType<typeof createResolvablePromise<T>>;
+
+class Subscription {
+  _unsubscribe: () => void = () => {};
+  unsubscribed = false;
+
+  value: RawCoMap | undefined;
+  status: "unknown" | "loading" | "loaded" | "unauthorized" | "unavailable" =
+    "unknown";
+
+  constructor(
+    public node: LocalNode,
+    public id: ID<CoMapSchema>,
+    public listener: (value: RawCoMap) => void,
+  ) {
+    const value = this.node.coValuesStore.get(this.id as any);
+
+    if (value.state.type === "available") {
+      this.status = "loaded";
+      this.subscribe(value.state.coValue.getCurrentContent() as RawCoMap);
+    } else {
+      this.status = "loading";
+      this.node.load(this.id as any).then((value) => {
+        if (this.unsubscribed) return;
+        // TODO handle the error states which should be transitive
+        if (value !== "unavailable") {
+          this.status = "loaded";
+          this.subscribe(value as RawCoMap);
+        } else {
+          this.status = "unavailable";
+        }
+      });
+    }
+  }
+
+  subscribe(value: RawCoMap) {
+    if (this.unsubscribed) return;
+
+    this._unsubscribe = value.subscribe((value) => {
+      this.listener(value);
+    });
+    this.listener(value);
+  }
+
+  unsubscribe() {
+    if (this.unsubscribed) return;
+    this.unsubscribed = true;
+    this._unsubscribe();
+  }
+}
+
+export class CoValueResolutionNode<S extends CoMapSchema> {
```

Comment:
This is use to build the resolution tree.

It tracks all the updates on the value and the linked childs and triggers the $update functions accordingly.